### ISSUE TITLE
False negative for `str` and `None` operands

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#3792]
**Describe the bug**

```py
from __future__ import annotations

def foo(a: str | None) -> None:
    foo(a=a + "foo")

foo(None)
```

**Expected behavior**

Pyright should flag the above code at line 3 as it will generate an exception at runtime,

**VS Code extension or command-line**

VS Code extension & command-line 1.1.265

**Comment by erictraut** [comment]
Thanks for the bug report. This turned out to be a pretty bad bug. An object that tracks the subtypes within a union is assumed to be immutable, but it was being modified. The subtypes were being reordered while iterating over them. Yikes! This bug has been present for several months. I'm surprised it didn't have other obvious side effects.

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.266, which I just published. It will also be included in a future release of pylance.
